### PR TITLE
[Refactor] Correct behaviour of CSWAP constructor

### LINKED ISSUE #180: [Bug] CSWAP control qubit definition
When I try to execute this lines:

```python3

import torch 
import random
from pyqtorch.circuit import QuantumCircuit
from pyqtorch.utils import random_state
from pyqtorch.primitive import CSWAP

n_qubits = torch.randint(low=2, high = 6, size=(1,)).item()
batch_size = torch.randint(low=1, high = 4, size=(1,)).item()
target = random.choice([i for i in range(n_qubits)])
control = random.choice([i for i in range(n_qubits) if i!=target])
input_state = random_state(n_qubits, batch_size)
cswap = CSWAP(target,control)
circ = QuantumCircuit(n_qubits, [cswap])
output_state = circ(input_state)
```

The output is:
```bash
  File "Pyqtorch/CSWAP.py", line 14, in <module>
    output_state = circ(input_state)
                   ^^^^^^^^^^^^^^^^^
  File "/Learn/Pyqtorch/Pyq_venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "Learn/Pyqtorch/Pyq_venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pyqtorch/circuit.py", line 73, in forward
    state = op(state, values)
            ^^^^^^^^^^^^^^^^^
  File "Learn/Pyqtorch/Pyq_venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "Learn/Pyqtorch/Pyq_venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/pyqtorch/primitive.py", line 60, in forward
    return apply_operator(
           ^^^^^^^^^^^^^^^
  File "pyqtorch/apply.py", line 49, in apply_operator
    operator = operator.view([2] * n_support * 2 + [operator.size(-1)])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: shape '[2, 2, 2, 2, 1]' is invalid for input of size 64
```

this error occurs only for the `CSWAP` gate and not for the other controlled gate like `CNOT`, `CY`, `CZ`. 
That's why, I think it's a bug link to the `CSWAP` only.

### LINKED ISSUE #184: [Feature] Modify the gates expra_repr method
Modify the `expra_repr` method to improve its clarity.

For instance: 
```python3
print(CNOT(control = (0,2), target = 1)
```
was returning:
```python3
CNOT((0,2,1))
```
now returns:
```python3
CNOT(control=(0,2), target=(1,))
```

### PR DESCRIPTION
**Pull request by EthanObadia**
- [x] Update `CSWAP` gate parameter requirement to only accept tuples as target, not integers, and added error handling for clarity.
- [x] Change the gates `extra_repr` method for more clarity .
- [x]  Check the behavior in `Qadence` after the modifcation of `extra_repr`

Closes #180.
Closes #184.

### CODE REVIEW

File: pyqtorch/primitive.py [hunk 1]
```diff
@@ -147,7 +147,9 @@ def __init__(self, control: int, target: int):
 
 
 class CSWAP(Primitive):
-    def __init__(self, control: int | tuple[int, ...], target: int):
+    def __init__(self, control: tuple[int, ...], target: int):
```
Comment (dominikandreasseitz) [inline pyqtorch/primitive.py:150]: the target is always a tuples since SWAP acts on two qubits. the control however can be int or tuple[int,..]

**Comment by dominikandreasseitz** [comment]
`SWAP`  is a two-qubit gate, hence has two "targets". 

atm in pyq, if you want to create a controlled version of that:

you do:
```
CSWAP(control=(0,1), target = 2)
```

however since we have several targets and possibly only one control, it would make more sense to me to do:
```
CSWAP(control=0, target = (1,2))
```



wdyt @jpmoutinho ?

**Comment by dominikandreasseitz** [comment]
in qadence we do:

```
class CSWAP(ControlBlock):
    """The CSWAP (Control-SWAP) gate."""

    name = OpName.CSWAP

    def __init__(self, control: int | tuple[int, ...], target1: int, target2: int) -> None:
```

File: pyqtorch/primitive.py (hunk 1 above)
Comment (EthanObadia) [reply pyqtorch/primitive.py:150]: So this version is correct right @jpmoutinho: 
```python3
class CSWAP(Primitive):
    def __init__(self, control: int | tuple[int, ...], target: tuple[int, ...]):
        if not isinstance(target, tuple) or len(target) == 2:
            raise ValueError("Target qubits must be a tuple with two qubits")
        super().__init__(OPERATIONS_DICT["CSWAP"], target)
        self.control = (control,) if isinstance(control, int) else control
        self.target = target
        self.qubit_support = self.control + (target,)
```

File: tests/test_circuit.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def test_differentiate_circuit(dtype: torch.dtype, batch_size: int, n_qubits: in
         pyq.Y(1),
         pyq.RX(0, "theta_0"),
         pyq.PHASE(0, "theta_1"),
-        pyq.CSWAP((0, 1), 2),
+        pyq.CSWAP(2, (0, 1)),
```
Comment (jpmoutinho) [inline tests/test_circuit.py:67]:
```suggestion
        pyq.CSWAP(0, (1, 2)),
```

File: tests/test_digital.py [hunk 3]
```diff
@@ -150,12 +151,12 @@ def test_CRY_state01_controlqubit_0() -> None:
 
 
 def test_CSWAP_state101_controlqubit_0() -> None:
-    result: Tensor = pyq.CSWAP((0, 1), 2)(product_state("101"), None)
-    assert torch.allclose(product_state("110"), result)
+    result: Tensor = pyq.CSWAP(2, (0, 1))(product_state("101"), None)
+    assert torch.allclose(product_state("011"), result)
```
Comment (jpmoutinho) [inline tests/test_digital.py:155]:
```suggestion
    result: Tensor = pyq.CSWAP(0, (1, 2))(product_state("101"), None)
    assert torch.allclose(product_state("110"), result)
```

File: tests/test_digital.py [hunk 4]
```diff
@@ -165,12 +166,12 @@ def test_CSWAP_state110_controlqubit_0() -> None:
...
         (state_100, state_100),
-        (state_101, state_110),
-        (state_110, state_101),
+        (state_101, state_011),
+        (state_110, state_110),
     ],
 )
 def test_CSWAP_controlqubits0(initial_state: Tensor, expected_state: Tensor) -> None:
-    cswap = pyq.CSWAP((0, 1), 2)
+    cswap = pyq.CSWAP(2, (0, 1))
```
Comment (jpmoutinho) [inline tests/test_digital.py:174]: Same as above, best to make the SWAP just `(0, (1, 2))` and fix the tests accordingly. It's a minor thing but just makes it easier to read 😅

File: tests/test_digital.py [hunk 5]
```diff
@@ -150,12 +151,12 @@ def test_CRY_state01_controlqubit_0() -> None:
...
 def test_CSWAP_state101_controlqubit_0() -> None:
-    result: Tensor = pyq.CSWAP((0, 1), 2)(product_state("101"), None)
-    assert torch.allclose(product_state("110"), result)
+    result: Tensor = pyq.CSWAP(2, (0, 1))(product_state("101"), None)
+    assert torch.allclose(product_state("011"), result)
 
 
 def test_CSWAP_state110_controlqubit_0() -> None:
-    result: Tensor = pyq.CSWAP((0, 1), 2)(product_state("101"), None)
+    result: Tensor = pyq.CSWAP(2, (0, 1))(product_state("110"), None)
```
Comment (jpmoutinho) [inline tests/test_digital.py:159]: This test can just be removed, same as the one before it. Just keep the one parameterized over various input, output states.

**Review by jpmoutinho — APPROVED** [review]
Thanks, indeed `control = i, target = (j, k)` is the expected syntax.

File: tests/test_digital.py (hunk 3 above)
Comment (EthanObadia) [reply tests/test_digital.py:155]: Removed the test.

File: tests/test_digital.py (hunk 5 above)
Comment (EthanObadia) [reply tests/test_digital.py:159]: Removed.

File: tests/test_digital.py (hunk 4 above)
Comment (EthanObadia) [reply tests/test_digital.py:174]: Modified.

File: pyqtorch/parametric.py [hunk 6]
```diff
@@ -112,7 +112,7 @@ def __init__(
     ):
         self.control = control if isinstance(control, tuple) else (control,)
         super().__init__(gate, target, param_name)
-        self.qubit_support = self.control + (self.target,)
+        self.qubit_support = self.control + (self.target,)  # type: ignore
```
[the remark this answers is not in the document]
Comment (dominikandreasseitz) [reply pyqtorch/parametric.py:115]: please use `# type: ignore[union-attr]`. its more verbose

[the remark this answers is not in the document]
Comment (EthanObadia) [reply pyqtorch/parametric.py:115]: While replacing, I encountered this error:
```bash
pyqtorch/parametric.py:115: error: Unsupported operand types for + ("tuple[int, ...]" and "tuple[int | tuple[int, ...]]")  [operator]
pyqtorch/parametric.py:115: note: Error code "operator" not covered by "type: ignore" comment
```

Thus, I used `# type: ignore[operator]`

**Review by dominikandreasseitz — APPROVED** [review]
thank you @EthanObadia , approved

**Pull request merged**